api/manifest.go

```diff
@@ -40,58 +38,98 @@ func apiManifestHandler(w http.ResponseWriter, r *http.Request) {
+			return getManifest(log, manifestAPI, string(latestSHA), paths)
 		}
-		body = cached.Value
+		log.Debugf("Latest SHA not found in cache: %v", err)
 	} else {
-		fetchedSHA, body, err = manifestAPI.GetManifestForSHA(sha)
+		// Attempt to find the manifest for a specific SHA in cache.
+		var err error
+		body, err = readByKey(mc, sha)
 		if err != nil {
-			return fetchedSHA, nil, err
-		}
-		if paths != nil {
-			if body, err = manifest.Filter(body, paths); err != nil {
-				return fetchedSHA, nil, err
-			}
+			log.Debugf("Manifest for SHA %s not found in cache: %v", sha, err)
 		}
+		// Do not return here yet as we still need to filter by paths.
 	}
 
-	item := &memcache.Item{
-		Key:   manifestCacheKey(fetchedSHA),
-		Value: body,
+	var fetchedSHA string
+	if body != nil {
+		// Cache hit; fetchedSHA is requested SHA, which is guaranteed to be specific.
+		fetchedSHA = sha
+	} else {
+		// Cache missed; download the manifest for real.
+		var err error
+		if fetchedSHA, body, err = manifestAPI.GetManifestForSHA(sha); err != nil {
+			return fetchedSHA, nil, err
+		}
+
+		// Write manifest to cache.
+		if err := writeByKey(mc, fetchedSHA, body); err != nil {
+			log.Errorf("Error writing manifest to cache: %v", err)
+		}
 	}
-	memcache.Set(ctx, item)
 
-	// Shorter expiry for latest SHA, to keep it current.
+	// Write latest SHA to cache, if needed.
 	if shared.IsLatest(sha) {
-		latestSHAItem := &memcache.Item{
-			Key:        manifestCacheKey("latest"),
-			Value:      []byte(fetchedSHA),
-			Expiration: time.Minute * 5,
+		if err := writeByKey(latestMC, "latest", []byte(fetchedSHA)); err != nil {
+			log.Errorf("Error writing latest SHA to cache: %v", err)
 		}
-		memcache.Set(ctx, latestSHAItem)
 	}
 
+	// Decompress the manifest and filter it by paths if needed.
 	gzReader, err := gzip.NewReader(bytes.NewReader(body))
 	if err != nil {
 		return fetchedSHA, nil, err
 	}
-	body, err = ioutil.ReadAll(gzReader)
-	return fetchedSHA, body, err
+	unzipped, err := ioutil.ReadAll(gzReader)
+	if err != nil {
+		return fetchedSHA, nil, err
+	}
+	if paths != nil {
+		var err error
+		if unzipped, err = manifest.Filter(unzipped, paths); err != nil {
+			return fetchedSHA, nil, err
+		}
+	}
+	return fetchedSHA, unzipped, err
 }
 
 func manifestCacheKey(sha string) string {
 	return fmt.Sprintf("MANIFEST-%s", sha)
 }
+
+func readByKey(readable shared.Readable, key string) ([]byte, error) {
+	ikey := manifestCacheKey(key)
+	reader, err := readable.NewReadCloser(ikey)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.ReadAll(reader)
+}
+
+func writeByKey(writable shared.ReadWritable, key string, body []byte) error {
+	ikey := manifestCacheKey(key)
+	writer, err := writable.NewWriteCloser(ikey)
+	if err != nil {
+		return err
+	}
+	n, err := writer.Write(body)
+	if err != nil {
+		return err
+	}
+	if n != len(body) {
+		return fmt.Errorf("incomplete write; expected %d bytes but %d written", len(body), n)
```

Comment:
```suggestion
		return fmt.Errorf("Incomplete write; expected %d bytes but %d written", len(body), n)
```

Reply:
Logs are capitalized, but errors are not. https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
Aha thank you. I will keep that in mind